chainer/links/loss/hierarchical_softmax.py

```diff
@@ -345,6 +391,78 @@ def create_huffman_tree(word_counts):
 
         return q.get()[2]
 
+    def sampling(self, x):
+        """Sampling word id for given input from tree path.
+
+        Args:
+            x (~chainer.Variable): Input variable for sampling word ids.
+                                 : Examples: Variable(
+                                            [
+                                              [0.2, 0.2, 0.3],
+                                              [0.1, 0.3, 0.1],
+                                              [0.2, 0.3, 0.4]
+                                            ])
+        Returns:
+            ~chainer.Variable: List of word indexes.
+                             : Examples: [0, 10, 3]
+        """
+        if len(self.tree) == 0:
+            raise ValueError('Empty tree')
+
+        xp = self.xp
+        parent2child = xp.array(self._func.parent2child).astype('i')
+        batchsize = x.data.shape[0]
+        start_ids = xp.zeros((batchsize, )).astype('i')
+        _lst_next_ids = []
+        _lst_choose_ids = []
+
+        def _sigmoid(x):
+            half = x.dtype.type(0.5)
+            return self.xp.tanh(x * half) * half + half
+
+        while True:
+            w = self.W.data[start_ids]
+            x_t = xp.transpose(x.data)
+            score = xp.sum(xp.dot(w, x_t), axis=1)
+            prob_left = _sigmoid(score)
+            prob_left = xp.reshape(prob_left, (batchsize, 1))
+            prob_right = xp.ones(prob_left.shape) - prob_left
+            prob = xp.concatenate([prob_left, prob_right], axis=1)
+
+            choosed_idx = xp.argmax(xp.random.gumbel(size=prob.shape) + prob,
+                                    axis=1)
+            rows = six.moves.range(batchsize)
+            rows = xp.array(rows)
+            columns = choosed_idx
+
+            nodes_ids = parent2child[start_ids]
+            next_ids = nodes_ids[rows, columns]
+            next_ids = xp.where(next_ids != LEAF, next_ids, FINISH_SAMPLING)
+
+            _lst_next_ids.append(start_ids)
+            _lst_choose_ids.append(choosed_idx)
+
+            # check whether all nodes are LEAF.
+            if xp.all(next_ids == FINISH_SAMPLING):
+                # if all nodes will reach leaf, then finish sampling.
+                break
+            start_ids = next_ids
+
+        sampling = [[] for _ in six.moves.range(batchsize)]
+        for _next_ids, _choose_ids in zip(_lst_next_ids, _lst_choose_ids):
+            for m in six.moves.range(batchsize):
+                if _next_ids[m] != FINISH_SAMPLING:
```

Comment:
Calling `__getitem__` for cupy.nadarray is slow. Convert cupy.ndarray to numpy.ndarray.